File: app/src/main/java/org/mifos/mobilebanking/ui/adapters/ShareAccountsListAdapter.java
package org.mifos.mobilebanking.ui.adapters;

import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.injection.ActivityContext;
import org.mifos.mobilebanking.models.accounts.share.ShareAccount;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ShareAccountsListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private final Context context;
    private List<ShareAccount> shareAccountsList = new ArrayList<>();

    @Inject
    public ShareAccountsListAdapter(@ActivityContext Context context) {
        this.context = context;
    }

    public void setShareAccountsList(List<ShareAccount> shareAccountsList) {
        this.shareAccountsList = shareAccountsList;
        notifyDataSetChanged();
    }

    public ShareAccount getItem(int position) {
        return shareAccountsList.get(position);
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder vh;
        View v = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.row_share_account, parent, false);
        vh = new ViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof ViewHolder) {

            ShareAccount shareAccount = getItem(position);
            ((ViewHolder) holder).tvClientShareAccountsNumber
                    .setText(shareAccount.getAccountNo());
            ((ViewHolder) holder).tvShareAccountsProductName
                    .setText(shareAccount.getProductName());
            ((ViewHolder) holder).llAccountDetail.setVisibility(View.GONE);

            if (shareAccount.getStatus().getActive()) {

                ((ViewHolder) holder).ivStatusIndicator.setColorFilter(ContextCompat.
                        getColor(context, R.color.deposit_green));
                setSharingAccountDetail(((ViewHolder) holder), shareAccount);

            } else if (shareAccount.getStatus().getApproved()) {

                ((ViewHolder) holder).ivStatusIndicator.setColorFilter(ContextCompat.
                        getColor(context, R.color.light_green));

            } else if (shareAccount.getStatus().getSubmittedAndPendingApproval()) {

                ((ViewHolder) holder).ivStatusIndicator.setColorFilter(ContextCompat.
                        getColor(context, R.color.light_yellow));

            } else {

                ((ViewHolder) holder).ivStatusIndicator.setColorFilter(ContextCompat.
                        getColor(context, R.color.light_blue));

            }
        }
    }

    private void setSharingAccountDetail(ViewHolder viewHolder, ShareAccount shareAccount) {
        viewHolder.llAccountDetail.setVisibility(View.VISIBLE);
        viewHolder.tvSharesPending.setText(String.valueOf(shareAccount
                .getTotalPendingForApprovalShares()));
        viewHolder.tvSharesApproved.setText(String.valueOf(shareAccount
                .getTotalApprovedShares()));
    }
    @Override
    public int getItemCount() {
        return shareAccountsList.size();
    }

    private LayerDrawable setCircularBackground(int colorId) {
        Drawable color = new ColorDrawable(ContextCompat.getColor(context, colorId));
        Drawable image = ContextCompat.getDrawable(context, R.drawable.circular_background);
        LayerDrawable ld = new LayerDrawable(new Drawable[]{image, color});
        return ld;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.tv_clientSharingAccountNumber)
        TextView tvClientShareAccountsNumber;

        @BindView(R.id.tv_shareAccountProductName)
        TextView tvShareAccountsProductName;

        @BindView(R.id.iv_status_indicator)
        AppCompatImageView ivStatusIndicator;

        @BindView(R.id.ll_account_detail)
        LinearLayout llAccountDetail;

        @BindView(R.id.tv_shares_pending)
        TextView tvSharesPending;

        @BindView(R.id.tv_shares_approved)
        TextView tvSharesApproved;

        public ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
        }
    }
}

File: app/src/main/java/org/mifos/mobile/ui/adapters/ShareAccountsListAdapter.kt
package org.mifos.mobile.ui.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import org.mifos.mobile.R
import org.mifos.mobile.core.model.entity.accounts.share.ShareAccount
import org.mifos.mobile.databinding.RowShareAccountBinding

class ShareAccountsListAdapter(
    private val onItemClick: (itemPosition: Int) -> Unit,
) : RecyclerView.Adapter<ShareAccountsListAdapter.ViewHolder>() {

    private var shareAccountsList: List<ShareAccount?>? = ArrayList()
    fun setShareAccountsList(shareAccountsList: List<ShareAccount?>?) {
        if (shareAccountsList != null) {
            this.shareAccountsList = shareAccountsList
        }
        notifyDataSetChanged()
    }

    fun getItem(position: Int): ShareAccount? {
        return shareAccountsList?.get(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding =
            RowShareAccountBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val shareAccount = getItem(position)
        holder.bind(shareAccount)
    }

    override fun getItemCount(): Int {
        if (shareAccountsList != null) {
            return shareAccountsList!!.size
        }
        return 0
    }

    inner class ViewHolder(private val binding: RowShareAccountBinding) :
        RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.setOnClickListener {
                onItemClick(bindingAdapterPosition)
            }
        }

        fun bind(shareAccount: ShareAccount?) {
            with(binding) {
                tvClientSharingAccountNumber.text = shareAccount?.accountNo
                tvShareAccountProductName.text = shareAccount?.productName
                llAccountDetail.visibility = View.GONE
                val context = itemView.context
                when {
                    shareAccount?.status?.active == true -> {
                        ivStatusIndicator.setColorFilter(
                            ContextCompat.getColor(
                                context,
                                R.color.deposit_green,
                            ),
                        )
                        setSharingAccountDetail(shareAccount)
                    }

                    shareAccount?.status?.approved == true -> {
                        ivStatusIndicator.setColorFilter(
                            ContextCompat.getColor(
                                context,
                                R.color.light_green,
                            ),
                        )
                    }

                    shareAccount?.status?.submittedAndPendingApproval == true -> {
                        ivStatusIndicator.setColorFilter(
                            ContextCompat.getColor(
                                context,
                                R.color.light_yellow,
                            ),
                        )
                    }

                    else -> {
                        ivStatusIndicator.setColorFilter(
                            ContextCompat.getColor(
                                context,
                                R.color.light_blue,
                            ),
                        )
                    }
                }
            }
        }

        private fun setSharingAccountDetail(shareAccount: ShareAccount) {
            with(binding) {
                llAccountDetail.visibility = View.VISIBLE
                tvSharesPending.text = shareAccount
                    .totalPendingForApprovalShares.toString()
                tvSharesApproved.text = shareAccount
                    .totalApprovedShares.toString()
            }
        }
    }
}
